README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# scico <img src="man/figures/logo.png" align="right" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/thomasp85/scico/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/thomasp85/scico/actions/workflows/R-CMD-check.yaml)
[![CRAN_Release_Badge](http://www.r-pkg.org/badges/version-ago/scico)](https://CRAN.R-project.org/package=scico)
[![Codecov test
coverage](https://codecov.io/gh/thomasp85/scico/branch/main/graph/badge.svg)](https://app.codecov.io/gh/thomasp85/scico?branch=main)
<!-- badges: end -->

This is a small package to provide access to the colour palettes
developed by Fabio Crameri and published at
<http://www.fabiocrameri.ch/colourmaps.php>. It uses more or less the
same api as [`viridis`](https://github.com/sjmgarnier/viridis) and
provides scales for [`ggplot2`](https://github.com/tidyverse/ggplot2)
without requiring `ggplot2` to be installed.

## Installation

`scico` can be installed from CRAN with `install.packages('scico')`. If
you want the development version then install directly from GitHub:

``` r
# install.packages("devtools")
devtools::install_github("thomasp85/scico")
```

## Palettes

`scico` provides 39 different palettes, all of which are perceptually
uniform and colourblind safe. An overview can be had with the
`scico_palette_show()` function:

``` r
library(scico)

scico_palette_show()
```

![](man/figures/README-unnamed-chunk-3-1.png)<!-- -->

Once you’ve decided on a palette you can generate colour values using
the `scico()` function:

``` r
scico(30, palette = 'lapaz')
#>  [1] "#190C64" "#1C176B" "#202272" "#212B79" "#243580" "#263D86" "#29478B"
#>  [8] "#2C5091" "#2F5996" "#33619A" "#37699D" "#3D71A0" "#4479A1" "#4D81A2"
#> [15] "#5688A4" "#608EA2" "#6B94A1" "#77999F" "#839E9C" "#90A198" "#9BA495"
#> [22] "#A9A895" "#B7AD96" "#C7B59C" "#D7BEA6" "#E5C9B3" "#F0D4C3" "#F7DFD3"
#> [29] "#FCE9E3" "#FEF2F2"
```

## ggplot2 support

`scico` provides relevant scales for use with `ggplot2`. It only
suggests `ggplot2` in order to stay lightweight, but if `ggplot2` is
available you’ll have access to the `scale_[colour|fill]_scico()`
functions:

``` r
library(ggplot2)
volcano <- data.frame(
  x = rep(seq_len(ncol(volcano)), each = nrow(volcano)),
  y = rep(seq_len(nrow(volcano)), ncol(volcano)),
  height = as.vector(volcano)
)
ggplot(volcano, aes(x = x, y = y, fill = height)) + 
  geom_raster() + 
  scale_fill_scico(palette = 'davos') 
```

![](man/figures/README-unnamed-chunk-5-1.png)<!-- -->

## References

- Crameri, Fabio. (2018, May 8). *Scientific colour maps (Version
  3.0.1)*. Zenodo. doi: 10.5281/zenodo.1243909
- Crameri, Fabio. (2018). *Geodynamic diagnostics, scientific
  visualisation and StagLab 3.0*. Geosci. Model Dev. Discuss. doi:
  10.5194/gmd-2017-328


NEWS.md
# scico (development version)

# scico 1.5.0

* Update palettes to Scientific Color Maps 8.0

# scico 1.4.0

* Add support for categorical palettes

# scico 1.3.1

* general upkeep

# scico 1.3.0

* Added new colour palettes (@michaelgrund, #11)
* Added midpoint argument to ggplot2 scales

# scico 1.2.0

* Added cyclic colour palettes

# scico 1.1.0

* Added discrete versions of the scales

# scico 1.0.0

* First release


tests/testthat.R
library(testthat)
library(scico)

test_check("scico")


tests/testthat/test_ggplot2_scales.R
context('test scale_color/fill_scico_d')



# Initial setup -----------------------------------------------------------

library(dplyr)
library(ggplot2)

base_color_plot <- mtcars %>% 
  mutate(cyl = factor(cyl)) %>% 
  ggplot(aes(x=wt, y=mpg)) + 
  geom_point(aes(color=cyl), size=10) 

base_fill_plot <- mtcars %>% 
  mutate(cyl = factor(cyl)) %>% 
  ggplot(aes(x=mpg)) + 
  geom_density(aes(fill=cyl)) 



# test scale_color_scico_d ------------------------------------------------

test_that('scale_color_scico_d is ScaleDiscrete',{
  expect_s3_class(scale_fill_scico_d(), 'ScaleDiscrete')
})

test_that('scale_color_scico_d returns a ggplot',{
  expect_s3_class(base_color_plot + scale_color_scico_d(), 'gg')
})

test_that('scale_color_scico_d takes palette args',{
  expect_s3_class(base_color_plot + 
                    scale_color_scico_d(palette = 'oslo'), 
                  'gg')
  expect_s3_class(base_color_plot + 
                    scale_color_scico_d(alpha = .5, begin = .5, end = .9), 
                  'gg')
  expect_s3_class(base_color_plot + 
                    scale_color_scico_d(direction = -1), 
                  'gg')
})

test_that('scale_color_scico_d fails as expected',{
  expect_error(base_color_plot + scale_color_scico_d(palette = 'inferno'))
  
  expect_error(base_color_plot + scale_color_scico_d(alpha = 2))
})



# test scale_fill_scico_d ------------------------------------------------

test_that('scale_fill_scico_d is ScaleDiscrete',{
  expect_s3_class(scale_fill_scico_d(), 'ScaleDiscrete')
})

test_that('scale_fill_scico_d returns a ggplot',{
  expect_s3_class(base_fill_plot + scale_fill_scico_d(), 'gg')
})

test_that('scale_fill_scico_d takes palette args',{
  expect_s3_class(base_fill_plot + 
                    scale_fill_scico_d(palette = 'oslo'), 
                  'gg')
  expect_s3_class(base_fill_plot + 
                    scale_fill_scico_d(alpha = .5, begin = .5, end = .9), 
                  'gg')
  expect_s3_class(base_fill_plot + 
                    scale_fill_scico_d(direction = -1), 
                  'gg')
})

test_that('scale_fill_scico_d fails as expected',{
  expect_error(base_fill_plot + scale_fill_scico_d(palette = 'inferno'))
  
  expect_error(base_fill_plot + scale_fill_scico_d(alpha = 2))
})


R/ggplot2_scales.R
#' Scales to use for ggplot2
#' 
#' These functions provide the option to use the scico palettes along with the
#' `ggplot2` package. It goes without saying that it requires `ggplot2` to work.
#' 
#' @param ... Arguments to pass on to `ggplot2::scale_colour_gradientn()`, 
#' `ggplot2::scale_fill_gradientn()`, `ggplot2::ggplot2::discrete_scale()`
#' @param aesthetics Character string or vector of character strings listing the
#'   name(s) of the aesthetic(s) that this scale works with. This can be useful,
#'   for example, to apply colour settings to the colour and fill aesthetics at
#'   the same time, via aesthetics = c("colour", "fill").
#' @param midpoint A midpoint to center the scale on, used primarily for 
#'   diverging and multisequential scales
#' @inheritParams scico
#' 
#' @return A `ScaleContinuous` or `ScaleDiscrete` object that can be added to a
#'   `ggplot` object
#'   
#' @name ggplot2-scales
#' @rdname ggplot2-scales
#' 
#' @examples 
#' 
#' if (require('ggplot2')) {
#'   volcano <- data.frame(
#'     x = rep(seq_len(ncol(volcano)), each = nrow(volcano)),
#'     y = rep(seq_len(nrow(volcano)), ncol(volcano)),
#'     height = as.vector(volcano)
#'   )
#'   
#'   ggplot(volcano, aes(x = x, y = y, fill = height)) + 
#'     geom_raster() + 
#'     scale_fill_scico(palette = 'tokyo') 
#'     
#'   ggplot(iris, aes(x=Petal.Width, y=Petal.Length)) +
#'     geom_point(aes(color=Species), size=10) + 
#'     scale_colour_scico_d() 
#' }
#' 
NULL

# Copied from ggplot2
#' @importFrom scales rescale_mid
mid_rescaler <- function(mid) {
  function(x, to = c(0, 1), from = range(x, na.rm = TRUE)) {
    rescale_mid(x, to, from, mid)
  }
}

# Continuous scales -------------------------------------------------------

#' @rdname ggplot2-scales
#' @importFrom scales rescale
#' @export
#' 
scale_colour_scico <- function(..., alpha = NULL, begin = 0, end = 1, direction = 1, palette = "bilbao", midpoint = NA) {
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop('ggplot2 is required for this functionality', call. = FALSE)
  }
  rescaler <- if (is.na(midpoint)) rescale else mid_rescaler(midpoint)
  ggplot2::scale_colour_gradientn(colours = scico(256, alpha, begin, end, direction, palette), ..., rescaler = rescaler)
}
#' @rdname ggplot2-scales
#' @export
#' 
scale_color_scico <- scale_colour_scico
#' @rdname ggplot2-scales
#' @importFrom scales rescale
#' @export
#' 
scale_fill_scico <- function(..., alpha = NULL, begin = 0, end = 1, direction = 1, palette = "bilbao", midpoint = NA) {
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop('ggplot2 is required for this functionality', call. = FALSE)
  }
  rescaler <- if (is.na(midpoint)) rescale else mid_rescaler(midpoint)
  ggplot2::scale_fill_gradientn(colours = scico(256, alpha, begin, end, direction, palette), ..., rescaler = rescaler)
}



# Discrete scales ---------------------------------------------------------

#' @rdname ggplot2-scales
#' @export
#' 
scale_colour_scico_d <-
  function(...,
           alpha = 1,
           begin = 0,
           end = 1,
           direction = 1,
           palette = 'batlow',
           aesthetics = "colour") {
    
  
  # initial checks (direction and beg/end are already checked with scico::scico)
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop('ggplot2 is required for this functionality', call. = FALSE)
  }
  
  if (alpha < 0 | alpha > 1) stop('alpha must be in [0,1]')
  
  if (!palette %in% scico_palette_names()) stop('Need to pick a scico palette')
  
  ggplot2::discrete_scale(aesthetics,
                 "scico_d",
                 scico_discrete(alpha, begin, end, direction, palette),
                 ...)
}

#' @rdname ggplot2-scales
#' @export
#' 
scale_color_scico_d <- scale_colour_scico_d


#' @rdname ggplot2-scales
#' @export
#' 
scale_fill_scico_d <- function(...,
                               alpha = 1,
                               begin = 0,
                               end = 1,
                               direction = 1,
                               palette = 'batlow',
                               aesthetics = "fill") {
  
  # initial checks (direction and beg/end are already checked with scico::scico)
  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop('ggplot2 is required for this functionality', call. = FALSE)
  }
  
  if (alpha < 0 | alpha > 1) stop('alpha must be in [0,1]')
  
  if (!palette %in% scico_palette_names()) stop('Need to pick a scico palette')
  
  ggplot2::discrete_scale(aesthetics,
                 "scico_d",
                 scico_discrete(alpha, begin, end, direction, palette),
                 ...)
}


scico_discrete <- function(alpha, begin, end, direction, palette) {
  function(n) {
    scico::scico(n, alpha, begin, end, direction, palette)
  }
}

data-raw/palettes.R
curl::curl_download(
  "https://zenodo.org/record/8035877/files/ScientificColourMaps8.zip",
  "./data-raw/raw_palettes.zip"
)

unzip("./data-raw/raw_palettes.zip", exdir = "./data-raw/")

palettes <- list(
  .categorical = list()
)

palette_names <- list.dirs(
  "./data-raw/ScientificColourMaps8/", 
  full.names = FALSE, 
  recursive = FALSE
)

palette_names <- grep("^\\w", palette_names, value = TRUE)

for (name in palette_names) {
  standard_pal <- paste0("./data-raw/ScientificColourMaps8/", name, "/", name, ".txt")
  if (file.exists(standard_pal)) {
    palettes[[name]] <- read.table(standard_pal, sep = " ", col.names = c("r", "g", "b"))
  }
  categorical_pal <- paste0("./data-raw/ScientificColourMaps8/", name, "/CategoricalPalettes/", name, "S_HEX.txt")
  if (file.exists(categorical_pal)) {
    palettes$.categorical[[name]] <- read.table(categorical_pal, skip = 2, comment.char = "", col.names = c("r", "g", "b", "name", "hex"))$hex
  }
}

unlink("./data-raw/raw_palettes.zip")
unlink("./data-raw/ScientificColourMaps8/", recursive = TRUE)

usethis::use_data(palettes, overwrite = TRUE, internal = TRUE)


R/palette.R
#' Scientific colour map palettes
#' 
#' This function constructs palettes of the specified size based on the colour 
#' maps developed by Fabio Crameri. It follows the same API style as `viridis()`
#' from the `viridisLite` package so anyone familiar with this package can 
#' easily adapt to that.
#' 
#' @param n The number of colours to generate for the palette
#' @param alpha The opacity of the generated colours. If specified rgba values
#' will be generated. The default (`NULL`) will generate rgb values which 
#' corresponds to `alpha = 1`
#' @param begin,end The interval within the palette to sample colours from. 
#' Defaults to `0` and `1` respectively
#' @param direction Either `1` or `-1`. If `-1` the palette will be reversed
#' @param palette The name of the palette to sample from. See 
#' [scico_palette_names()] for a list of possible names
#' @param categorical Boolean. Should the categorical palettes be returned
#' 
#' @return A character vector of length `n` with hexencoded rgb(a) colour values
#' 
#' @references 
#' <http://www.fabiocrameri.ch/colourmaps.php>
#' 
#' Crameri, Fabio. (2021, September 12). *Scientific colour maps (Version 7.0.1)*. Zenodo. \doi{10.5281/zenodo.5501399}
#' Crameri, Fabio. (2018). *Geodynamic diagnostics, scientific visualisation and StagLab 3.0*. Geosci. Model Dev. Discuss. \doi{10.5194/gmd-2017-328}
#' 
#' @export
#' @importFrom grDevices rgb colorRamp
#' 
#' @examples 
#' 
#' # Use the default palette
#' scico(15)
#' 
#' # Flip the direction
#' scico(15, direction = -1)
#' 
#' # Take a subset of a palette
#' scico(15, begin = 0.3, end = 0.6, palette = 'berlin')
#' 
scico <- function(n, alpha = NULL, begin = 0, end = 1, direction = 1, palette = "bilbao", categorical = FALSE) {
  if (begin < 0 | begin > 1 | end < 0 | end > 1) {
    stop("begin and end must be in [0,1]")
  }
  
  if (abs(direction) != 1) {
    stop("direction must be 1 or -1")
  }
  
  if (direction == -1) {
    tmp <- begin
    begin <- end
    end <- tmp
  }
  
  palette <- scico_palette_data(palette, categorical = categorical)
  if (categorical) {
    if (n > 100) {
      stop("Cannot create categorical palettes with more than 100 colours")
    }
    return(palette[seq_len(n)])
  }
  
  map_cols <- rgb(palette$r, palette$g, palette$b)
  fn_cols <- colorRamp(map_cols, space = "Lab", interpolate = "spline")
  cols <- fn_cols(seq(begin, end, length.out = n))
  if (is.null(alpha)) {
    rgb(cols[, 1], cols[, 2], cols[, 3], maxColorValue = 255)
  } else {
    rgb(cols[, 1], cols[, 2], cols[, 3], alpha = alpha * 255, maxColorValue = 255)
  }
}
#' Access raw palette data
#' 
#' These functions gives access to the data underlying the palettes
#' 
#' @param categorical Boolean. Should the categorical palettes be returned
#' 
#' @keywords internal
#' 
#' @name scico_data
#' @rdname scico_data
#' 
#' @export
#' 
scico_palette_names <- function(categorical = FALSE) {
  if (categorical) {
    names(palettes$.categorical)
  } else {
    setdiff(names(palettes), ".categorical")
  }
}
#' @rdname scico_data
#' 
#' @param palette The name of the palette to fetch data for
#' @param categorical Boolean. Should the categorical palettes be returned
#' 
#' @export
#' 
scico_palette_data <- function(palette, categorical = FALSE) {
  palette <- match.arg(palette, scico_palette_names(categorical))
  if (categorical) {
    palettes$.categorical[[palette]]
  } else {
    palettes[[palette]]
  }
}

#' Show the different scico palettes
#' 
#' This is a simple function to show a gradient of the different palettes 
#' available in the `scico` package
#' 
#' @param palettes One or more palette names to show
#' @param categorical Boolean. Should the categorical palettes be returned
#' @param n How many colours should be shown
#' 
#' 
#' @importFrom grDevices n2mfrow
#' @importFrom graphics image par
#' @export
#' 
#' @examples 
#' 
#' scico_palette_show()
#' scico_palette_show(categorical = TRUE)
#' 
scico_palette_show <- function(palettes = scico_palette_names(categorical), categorical = FALSE, n = if (categorical) 6 else 100) {
  dims <- n2mfrow(length(palettes))
  oldpar <- par(mfrow = dims, mai = par('mai')/5)
  on.exit(par(oldpar))
  
  for (i in palettes) {
    image(matrix(seq_len(n), ncol = 1), col = scico(n, palette = i, categorical = categorical), main = i, axes = FALSE, )
  }
}
